File: ProblemSets/Week1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number = get_long("Number: ");
    int nrOfDigits = 0, sum = 0, tempDigit = 0, firstDigit = 0, secondDigit = 0;

    while (number != 0)
    {
        tempDigit = number % 10;

        if (number > 10 && number < 100)
        {
            secondDigit = tempDigit;
        }
        else if (number > 0 && number < 10)
        {
            firstDigit = tempDigit;
        }

        if (nrOfDigits % 2 != 0)
        {
            // If tempDigit > 9 add products digits,
            // (same as subtracting 9 from resulting product ex. 12 => 1+2=3 / 12-9=3)
            if (tempDigit * 2 > 9)
            {
                tempDigit = tempDigit * 2 - 9;
            }
            else
            {
                tempDigit *= 2;
            }
        }

        sum = sum + tempDigit;
        number /= 10;
        nrOfDigits++;
    }

    if (sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if ((firstDigit == 3 && (secondDigit == 4 || secondDigit == 7)) && nrOfDigits == 15)
    {
        printf("AMEX\n");
    }
    else if ((firstDigit == 5 &&
              (secondDigit == 1 || secondDigit == 2 || secondDigit == 3 || secondDigit == 4 || secondDigit == 5)) &&
             nrOfDigits == 16)
    {
        printf("MASTERCARD\n");
    }
    else if (firstDigit == 4 && (nrOfDigits == 13 || nrOfDigits == 16))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: ProblemSets/Week6/credit.py
from cs50 import get_string
import re

# Get input from user
n = get_string("Number: ")
sum = 0

# Check if user input contains only numbers and is of appropriate length
if re.fullmatch(r"\d+", n) and (len(n) >= 13 and len(n) <= 16):
    # Loop through user input in reverse order
    for i in range(len(n)):
        if i == 0:
            # Get last digit of string
            tempDigit = int(n[len(n) - 1])
        # Get every second digit starting from the second-to-last digit
        elif i % 2 == 1:
            tempDigit = int(n[len(n) - (i + 1)]) * 2

            # If tempDigit > 9 add products digits,
            # (same as subtracting 9 from resulting product ex. 12 => 1+2=3 / 12-9=3)
            if (tempDigit > 9):
                tempDigit -= 9
        # Get every second digit starting from the last digit
        else:
            tempDigit = int(n[len(n) - (i + 1)])

        sum += tempDigit
else:
    print("Input must be a numeric value between 13-16 numbers!")

if sum % 10 != 0:
    print("INVALID")
elif len(n) == 15 and (int(n[0]) == 3 and int(n[1]) in [4, 7]):
    print("AMEX")
elif len(n) == 16 and (int(n[0]) == 5 and int(n[1]) in [1, 2, 3, 4, 5]):
    print("MASTERCARD")
elif len(n) in [13, 16] and int(n[0]) == 4:
    print("VISA")
else:
    print("INVALID")
